examples/barotropic_instability.jl

```diff
@@ -0,0 +1,288 @@
+#
+# # The "power method" for calculating growth rates and eigenmodes
+#
+# An "instability" is a small-amplitude solution that develops due to the presence
+# of an unstable "basic state" or background flow. Because these solutions are
+# small amplitude, we express any of the prognostic variables as
+#
+# ```math
+# u = û exp(σ t)
+# ```
+#
+# where ``σ`` is the "growth rate" of the instability. Our object is to
+# compute ``σ`` and also to get a feel for ``û``, which represents the 
+# "eigenmode", or the spatial structure of the instability.
+#
+# The power method iteratively simulates the growth of the instability, using
+# a "rescaling" method to successively isolate the instability from other motions
+# that develop during a simulation of the fully nonlinear equations.
+#
+# For this we design a criterion for stopping a simulation based on the amplitude
+# of the perturbation ``u``-velocity field:
+
+# 
+# # The domain 
+
+using Oceananigans
+
+grid = RegularCartesianGrid(size=(64, 64, 1), x=(-2π, 2π), y=(-2, 5), z=(0, 1),
+                                  topology=(Periodic, Bounded, Bounded))
+
+# # The background flow
+#
+# Our background flow is a boundary-hugging "Bickley jet":
+
+using Oceananigans.Fields
+
+bickley_jet(x, y, z, t) = sech(y)^2
+
+U = BackgroundField(bickley_jet)
+
+# which looks like
+
+using Oceananigans.Grids
+using Plots
+
+y = ynodes(Cell, grid)
+
+background_flow = plot(bickley_jet.(0, y, 0, 0), y, label=nothing,
+                       linewidth=2, xlabel="U(y)", ylabel="y",
+                       title="Boundary-hugging Bickley jet")
+
+display(background_flow)
+
+# The Bickley jet is well studied, ie...
+# But we usually don't put it right on a boundary.
+#
+# # The ``β``-plane
+#
+# To season our problem with Oceananographic flavor, we investigate the instability
+# on the ``β``-plane, where the background rotation rate varies in ``y``:
+
+coriolis = BetaPlane(f₀=1, β=0.01)
+
+# Note that instability is suppressed if ``β > 1`` [cite].
+
+# # The model
+
+using Oceananigans.Advection
+
+model = IncompressibleModel(timestepper = :RungeKutta3, 
+                              advection = UpwindBiasedFifthOrder(),
+                                   grid = grid,
+                               coriolis = coriolis,
+                      background_fields = (u=U,),
+                                closure = IsotropicDiffusivity(ν=1e-6),
+                               buoyancy = nothing,
+                                tracers = nothing)
+
+# # A _Power_ful algorithm
+#
+# We set up an algorithm that rescales the velocity field whenever
+# the ``u``-component of the velocity field exeeds some threshold.
+
+velocity_exceeds_threshold(sim; threshold=1e-2) = maximum(abs, interior(sim.model.velocities.u)) > threshold
+
+simulation = Simulation(model, Δt=0.1, iteration_interval=100,
+                        progress=sim -> @info("Model iteration: $(sim.model.clock.iteration)"))
+
```

Comment:
I get a warning:

```julia
julia> simulation = Simulation(model, Δt=0.1, iteration_interval=100,
                               progress=sim -> @info("Model iteration: $(sim.model.clock.iteration)"))
┌ Warning: This simulation will run forever as stop iteration = stop time = wall time limit = Inf.
└ @ Oceananigans.Simulations ~/Research/Oceananigans.jl/src/Simulations/simulation.jl:63
```

 @glwagner  you mentioned that this is bogus warning... Should we fix it before this PR? Otherwise the warning will appear in the documentation.

Reply:
Yes, this warning should be moved to `run!(simulation)` with the appropriate modifications so that it is correct.